### ISSUE TITLE
Introduce Backbone.BaseRouter

### ISSUE THREAD

**Issue by jmeas** [marionettejs/backbone.marionette#2092]
The most common thing people want to do with the Backbone router is add more abstractions to it. Unlike other BB classes, though, the router is far more difficult to work with. I would go so far as to say that Backbone actively discourages modifications to the History or Router classes. These two classes are by far the most opinionated objects in Backbone, which is a shame because the router is so, so bad.

The common things people want to do are:
- Add before and after hooks to routes
- Redirect under certain conditions
- Security / auth

These things are tough to do with the regular Backbone router, but trivial to implement with the [BaseRouter](https://github.com/jmeas/backbone.base-router). The best part of the base router is that it can be used to reconstruct Backbone's Router or Marionette's AppRouter in a backwards compatible way. But it can also be used to create far more powerful routers, like [this one](https://github.com/jmeas/gistbook/blob/master/client.src/vendor/routing/router.js), which is inspired by Angular UI's ui-router and Ember's router.

With all of this said, there are a few caveats. The BaseRouter strips away the opinions Backbone gives to its router, and that's about it. But it doesn't change anything else fundamental about it. This is a problem.

One thing I'd love to see is named routing. Rather than:

`router.navigate('book/23/chapter/2');`

wouldn't it be great if we could do

`router.navigate('chapter', {book: 23, id: 2});`

From here, I'd love to introduce a handlerbars helper like @thejameskyle's link-to prototype that I've lost the link to.

These things might be possible with the BaseRouter, but I haven't put the time into researching it further.

**Comment by rbu** [comment]
Has this router been going into 3.0 or is it still planned to integrate this router?

**Comment by scott-w** [comment]
Hi @rbu, this was dropped from the v3.0 release so we could focus on getting the existing improvements out the door.

We're still discussing how to improve the `Router` for Marionette though we don't have a specific version in mind yet. We're open to feedback on where you think improvements could be made.

**Comment by paulfalgout** [comment]
Similar to my comment on https://github.com/marionettejs/backbone.marionette/issues/2280 I don't think Marionette needs to bundle a router.  We should loosely support a few, but I don't think Marionette needs to be in the router business.  The bundled one is backbone's which is an ok starting place.. And when you need to find a better solution we should have documentation, perhaps pointing to baserouter and others that would work. @marionettejs/marionette-core ?

**Comment by scott-w** [comment]
Sounds good. Do we have a list of routers somewhere that we can reference? I could try to write up a Routing document to go through the different strategies, recommending a good one for newcomers.

**Comment by paulfalgout** [comment]
Certainly nothing too definite. A very in depth analysis, though a wip and a little out of date is https://github.com/jmeas/router-comparison

**Comment by megawac** [comment]
Agreed, as long as we document a proposed migration strategy if we drop our
router

**Comment by jmeas** [comment]
I'm :+1: to dropping the router altogether from Marionette. Within routing docs, I would strongly discourage folks from using Backbone's router. My favorite feature about non-Backbone frameworks like React or Ember isn't the smart diffing (which is convenient, don't get me wrong), but their amazing routers.

Router suggestions:

[Stateman](https://github.com/leeluolee/stateman)
tl;dr: Angular-UI style routing, without the Angular

UI-Router is pretty decent, so this is also pretty decent. I'd use this on a large webapp myself. A few features are missing that I like, like a NotFound abstraction at any level (there's only a global one). You can get around this with regexp or the pathing DSL. Documentation is pretty thorough, but missing key things like explaining its pathing DSL. Also, it has a bus factor of 1, so I'd only use it if you plan to help maintain it.

[Cherrytree](https://github.com/QubitProducts/cherrytree)
Ember's router: [router.js](https://github.com/tildeio/router.js/)
tl;dr: Nested routes, like Ember, but without [the transition lifecycle](https://github.com/tildeio/router.js/#transition-callbacks).

This router has always struck me as odd, because the transition callbacks are one of my favorite part of routers. Well, you don't get those with cherrytree. In its place, you get Express-style middleware which seems to me seems substantially less useful, but some folks may prefer that.

I wouldn't use this router, because my understanding of their middleware-style routing makes it strictly less capable than Ember's routing algorithm. Which is a bummer, because otherwise everything about this router seems quite nice.

[router.js](https://github.com/tildeio/router.js)
tl;dr: Ember's router

This is what Ember's router is built on top of. It has all of the features needed for good routing, so I'd definitely use it. It's not perfect, though:

1. the code it's built on top of (including route-recognizer) is difficult to read and understand
2. the API can also be difficult to understand sometimes
3. the docs don't help provide clarity as to how to use it

It might take some messing around and looking at tests to figure out how to use it, but it's definitely a solid router.

---

No standalone router I've found comes up to the quality of the actual router built into Ember, or React Router, but these 3 are noteworthy mentions.

There are definitely others, too...these are just 3 that I've looked in depth at.

**Comment by blikblum** [comment]
Recently i've been looking into a router solution to integrate in my Marionette work since, IMO, none of the Backbone based routers provides the same level of developer experience like e.g. React Router.

I looked into cherrytree, router.js and [abstract state router](https://github.com/TehShrike/abstract-state-router).

I've got a similar impression of @jmeas about router.js regarding being difficult to understand. Also seems a bit heavily opinionated making harder to map to Marionette concepts

Abstract state router is easier to understand but requires a render adapter that is more tailored for virtual dom like libraries

I did not know stateman

In the first look, i've got the same impression about cherrytree being too much bare bones and discarded it. But in a revisit, looking carefully, i realized that it could handle well the job of being the engine to a router that integrates well with Marionette

So, five days ago, i started to work in a cherrytree based routing. While still in an Proof of Concept stage, i'm confident that i can get a router with similar feature level of the modern ones that integrates well with Marionette concepts

It's far from being ready for usage but if someone wants to take a look the code is [here](https://bitbucket.org/blikblum/sharemonitor/src/82d1d08906dfa22e8e8307268de3b82248b0bb8c/clients/ionic-mn/scripts/marionette.routing.js?at=master&fileviewer=file-view-default).

BTW: i'm all in favor of ditching Marionette.AppRouter altogether. After looking of what it provides, i preferred to keep with the standard Backbone.Router
